File: scripts/opcodes/u32_std.js
import './std.js'

export const u32_push = value => [
    (value & 0xff000000) >>> 24,
    (value & 0x00ff0000) >>> 16,
    (value & 0x0000ff00) >>> 8,
    (value & 0x000000ff),
]

export const u32_equalverify = [
    4,
    OP_ROLL,
    OP_EQUALVERIFY,
    3,
    OP_ROLL,
    OP_EQUALVERIFY,
    OP_ROT,
    OP_EQUALVERIFY,
    OP_EQUALVERIFY,
]

export const u32_equal = [
    4,
    OP_ROLL,
    OP_EQUAL, OP_TOALTSTACK,
    3,
    OP_ROLL,
    OP_EQUAL, OP_TOALTSTACK,
    OP_ROT,
    OP_EQUAL, OP_TOALTSTACK,
    OP_EQUAL,
    OP_FROMALTSTACK, OP_BOOLAND,
    OP_FROMALTSTACK, OP_BOOLAND,
    OP_FROMALTSTACK, OP_BOOLAND,
]

export const u32_notequal = [
    4,
    OP_ROLL,
    OP_EQUAL, OP_NOT, OP_TOALTSTACK,
    3,
    OP_ROLL,
    OP_EQUAL, OP_NOT, OP_TOALTSTACK,
    OP_ROT,
    OP_EQUAL, OP_NOT, OP_TOALTSTACK,
    OP_EQUAL, OP_NOT,
    OP_FROMALTSTACK, OP_BOOLOR,
    OP_FROMALTSTACK, OP_BOOLOR,
    OP_FROMALTSTACK, OP_BOOLOR,
]

export const u32_toaltstack = [
    OP_TOALTSTACK,
    OP_TOALTSTACK,
    OP_TOALTSTACK,
    OP_TOALTSTACK,
]

export const u32_fromaltstack = [
    OP_FROMALTSTACK,
    OP_FROMALTSTACK,
    OP_FROMALTSTACK,
    OP_FROMALTSTACK,
]

export const u32_drop = [
    OP_2DROP,
    OP_2DROP,
]

export const u32_roll = a => {
    a = (a + 1) * 4 - 1
    return [
        a, OP_ROLL,
        a, OP_ROLL,
        a, OP_ROLL,
        a, OP_ROLL,
    ]
}

export const u32_pick = a => {
    a = (a + 1) * 4 - 1
    return [
        a, OP_PICK,
        a, OP_PICK,
        a, OP_PICK,
        a, OP_PICK,
    ]
}


export const u32_dup = [
    3,
    OP_PICK,
    OP_TOALTSTACK,
    OP_3DUP,
    OP_FROMALTSTACK,
    OP_ROT,
    OP_ROT,
    OP_2SWAP,
    OP_SWAP,
    OP_2SWAP,
]


const u32_compress = [
    OP_SWAP,
    OP_ROT,
    3,
    OP_ROLL,
    OP_DUP,
    127,
    OP_GREATERTHAN,
    OP_IF,
        128,
        OP_SUB,
        1,
    OP_ELSE,
        0,
    OP_ENDIF,
    OP_TOALTSTACK,
    OP_256MUL,
    OP_ADD,
    OP_256MUL,
    OP_ADD,
    OP_256MUL,
    OP_ADD,
    OP_FROMALTSTACK,
    OP_IF,
        OP_NEGATE,
    OP_ENDIF,
]

File: src/u32/u32_std.rs
#![allow(dead_code)]

use crate::pseudo::{push_to_stack, OP_256MUL, OP_4DUP};

use crate::treepp::{script, Script};

/// Pushes a value as u32 element onto the stack
pub fn u32_push(value: u32) -> Script {
    script! {
        //optimization
        if (value >> 24 & 0xff) == (value >> 16 & 0xff) &&
            (value >> 24 & 0xff) == (value >> 8 & 0xff) &&
            (value >> 24 & 0xff) == (value & 0xff) {

                { push_to_stack((value >> 24 & 0xff) as usize,4) }
        }
        else{

                {value >> 24 & 0xff}
                {value >> 16 & 0xff}
                {value >>  8 & 0xff}
                {value & 0xff}
        }
    }
}

/// Marks transaction as invalid if the top two stack value are not equal
pub fn u32_equalverify() -> Script {
    script! {
        4
        OP_ROLL
        OP_EQUALVERIFY
        3
        OP_ROLL
        OP_EQUALVERIFY
        OP_ROT
        OP_EQUALVERIFY
        OP_EQUALVERIFY
    }
}
/// Returns 1 if the top two u32 are equal, 0 otherwise
pub fn u32_equal() -> Script {
    script! {
        4
        OP_ROLL
        OP_EQUAL OP_TOALTSTACK
        3
        OP_ROLL
        OP_EQUAL OP_TOALTSTACK
        OP_ROT
        OP_EQUAL OP_TOALTSTACK
        OP_EQUAL
        OP_FROMALTSTACK OP_BOOLAND
        OP_FROMALTSTACK OP_BOOLAND
        OP_FROMALTSTACK OP_BOOLAND
    }
}

/// Returns 1 if the top two u32 are not equal, 0 otherwise
pub fn u32_notequal() -> Script {
    script! {
        4
        OP_ROLL
        OP_NUMNOTEQUAL OP_TOALTSTACK
        3
        OP_ROLL
        OP_NUMNOTEQUAL OP_TOALTSTACK
        OP_ROT
        OP_NUMNOTEQUAL OP_TOALTSTACK
        OP_NUMNOTEQUAL
        OP_FROMALTSTACK OP_BOOLOR
        OP_FROMALTSTACK OP_BOOLOR
        OP_FROMALTSTACK OP_BOOLOR
    }
}

/// Puts the top u32 element onto the top of the alt stack. Removes it from the main stack.
pub fn u32_toaltstack() -> Script {
    script! {
        OP_TOALTSTACK
        OP_TOALTSTACK
        OP_TOALTSTACK
        OP_TOALTSTACK
    }
}

/// Puts the top u32 element of the alt stack onto the top of the main stack. Removes it from the alt stack.
pub fn u32_fromaltstack() -> Script {
    script! {
        OP_FROMALTSTACK
        OP_FROMALTSTACK
        OP_FROMALTSTACK
        OP_FROMALTSTACK
    }
}

/// Duplicates the top u32 stack element
pub fn u32_dup() -> Script {
    script! { OP_4DUP }
}

/// Removes the top u32 element from the stack.
pub fn u32_drop() -> Script {
    script! {
        OP_2DROP
        OP_2DROP
    }
}

/// The u32 element n back in the stack is moved to the top.
pub fn u32_roll(n: u32) -> Script {
    let n = (n + 1) * 4 - 1;
    script! {
        {n} OP_ROLL
        {n} OP_ROLL
        {n} OP_ROLL
        {n} OP_ROLL
    }
}

/// The u32 element n back in the stack is copied to the top.
pub fn u32_pick(n: u32) -> Script {
    let n = (n + 1) * 4 - 1;
    script! {
        {n} OP_PICK
        {n} OP_PICK
        {n} OP_PICK
        {n} OP_PICK
    }
}

/// The top u32 element is compressed into a single 4-byte word
pub fn u32_compress() -> Script {
    script! {
        OP_SWAP
        OP_ROT
        3
        OP_ROLL
        OP_DUP
        127
        OP_GREATERTHAN
        OP_IF
            128
            OP_SUB
            1
        OP_ELSE
            0
        OP_ENDIF
        OP_TOALTSTACK
        OP_256MUL
        OP_ADD
        OP_256MUL
        OP_ADD
        OP_256MUL
        OP_ADD
        OP_FROMALTSTACK
        OP_IF
            OP_NEGATE
        OP_ENDIF
    }
}

#[cfg(test)]
mod test {

    use crate::run;
    use crate::treepp::script;
    use crate::u32::u32_std::*;
    use rand::Rng;

    #[test]
    fn test_u32_push() {
        let script = script! {
            { u32_push(0x01020304) }
            0x04
            OP_EQUALVERIFY
            0x03
            OP_EQUALVERIFY
            0x02
            OP_EQUALVERIFY
            0x01
            OP_EQUAL
        };
        run(script);
    }

    #[test]
    fn test_with_u32_compress() {
        let mut rng = rand::thread_rng();

        for _ in 0..30 {
            let mut origin_value0: u32 = rng.gen();
            origin_value0 = (origin_value0 % 1) << 31;
            let mut origin_value1: u32 = rng.gen();
            origin_value1 = (origin_value1 % 1) << 31;

            let v = origin_value0 + origin_value1;

            let script = script! {
                { u32_push(origin_value0)}
                { u32_compress()}
                { u32_push(origin_value1)}
                { u32_compress()}
                OP_ADD
                { u32_push(v)}
                { u32_compress()}
                OP_EQUAL

            };

            run(script);
        }
    }
}
